libtock/alarm_timer.c

```diff
@@ -123,45 +243,89 @@ void alarm_cancel(alarm_t* alarm) {
 
   alarm->prev = NULL;
   alarm->next = NULL;
+}
 
+// The intermediate callback that handles overflows of alarm. This is used by
+// `timer_in` to handle cases where `alarm_at` would normally be unable to
+// support the whole timer length. `alarm_at` can only keep track of up to 2^32
+// ticks.
+static void overflow_callback(int                          last_timer_fire_time,
+                              __attribute__ ((unused)) int unused1,
+                              __attribute__ ((unused)) int unused2,
+                              void*                        overflow_ud) {
+  tock_timer_t* tock_timer = (tock_timer_t*)overflow_ud;
+
+  if (tock_timer->overflows_left == 0) {
+    // no overflows left, schedule last alarm with original callback
+    alarm_at(
+      last_timer_fire_time,
+      tock_timer->remaining_ticks,
```

Comment:
If this alarm ends up set in the past (ie because remaining ticks happens to be its lowest possible value, and it took awhile for the kernel to get around to actually running this upcall), is everything ok?

Reply:
In this case, the upcall should fire immediately once the kernel gets around to scheduling it. We think this is the behavior based on [this comment and implementation in the alarm capsule](https://github.com/tock/tock/blob/master/capsules/core/src/alarm.rs#L91-L107).

For this case where we're setting the last alarm that won't overflow, this will be fine. The user's callback will get scheduled as soon as possible due to the behavior of the alarm capsule.